src/ModelContextProtocol/Client/AutoDetectingClientSessionTransport.cs

```diff
@@ -0,0 +1,200 @@
+    /// </summary>
+    internal ITransport? ActiveTransport => _streamableHttpTransport != null ? (ITransport)_streamableHttpTransport : _sseTransport;
+
+    public ChannelReader<JsonRpcMessage> MessageReader => _delegatingChannelReader;
+
+    /// <inheritdoc/>
+    public async Task SendMessageAsync(JsonRpcMessage message, CancellationToken cancellationToken = default)
+    {
+        if (_streamableHttpTransport == null && _sseTransport == null)
+        {
+            var rpcRequest = message as JsonRpcRequest;
+            
+            // The first message must be an initialize request
+            Debug.Assert(rpcRequest != null && rpcRequest.Method == RequestMethods.Initialize, 
+                "First message must be an initialize request");
+
+            // Try StreamableHttp first
+            _streamableHttpTransport = new StreamableHttpClientSessionTransport(_options, _httpClient, _loggerFactory, _name);
+            
+            try
+            {
+                LogAttemptingStreamableHttp(_name);
+                var response = await _streamableHttpTransport.SendInitialRequestAsync(message, cancellationToken).ConfigureAwait(false);
+                
+                // If the status code is not success, fall back to SSE
+                if (!response.IsSuccessStatusCode)
+                {
+                    LogStreamableHttpFailed(_name, response.StatusCode);
+                    
+                    try
+                    {
+                        await _streamableHttpTransport.DisposeAsync().ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        _streamableHttpTransport = null;
+                    }
+                    
+                    await InitializeSseTransportAsync(message, cancellationToken).ConfigureAwait(false);
+                    return;
+                }
+                
+                // Process the streamable HTTP response using the transport
+                await _streamableHttpTransport.SendMessageAsync(message, cancellationToken).ConfigureAwait(false);
+                
+                // Signal that we have established a connection
+                LogUsingStreamableHttp(_name);
+                _delegatingChannelReader.SetConnected();
+            }
+            catch (Exception ex)
+            {
+                LogStreamableHttpException(_name, ex);
+                
+                try
+                {
+                    if (_streamableHttpTransport != null)
+                    {
+                        await _streamableHttpTransport.DisposeAsync().ConfigureAwait(false);
+                    }
+                }
+                finally
+                {
+                    _streamableHttpTransport = null;
+                }
+                
+                await InitializeSseTransportAsync(message, cancellationToken).ConfigureAwait(false);
```

Comment:
We should only call InitializeSseTransportAsync when SendInitialRequestAsync doesn't throw but does return an unsuccessful status code.